<a href="https://colab.research.google.com/github/trellums/DS-Unit-1-Sprint-2-Statistics/blob/master/mod2_sprint2_dantell_LS_Chi2_Tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Assignment - Practice Chi^2 Tests

Use the following dataset relating to math scores of students in two different Portugese schools:

<https://archive.ics.uci.edu/ml/datasets/Student+Performance>

### 1) Load the dataset specifically relating to math scores as a new dataframe.

There are two datasets in the `student.zip` file, make sure you use `student-mat.csv`.


In [ ]:
# YOUR WORK HERE 

In [ ]:
import pandas as pd 
import matplotlib.pyplot as plt
import numpy as np
from scipy import stats

In [ ]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip

--2020-07-16 16:31:20--  https://archive.ics.uci.edu/ml/machine-learning-databases/00320/student.zip
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20478 (20K) [application/x-httpd-php]
Saving to: ‘student.zip’

student.zip         100%[===================>]  20.00K  --.-KB/s    in 0.03s   

2020-07-16 16:31:20 (709 KB/s) - ‘student.zip’ saved [20478/20478]



In [ ]:
!unzip student.zip

Archive:  student.zip
  inflating: student-mat.csv         
  inflating: student-por.csv         
  inflating: student-merge.R         
  inflating: student.txt             


In [ ]:
df=pd.read_csv('student-mat.csv',sep=';')
print(df.shape)
df.head()

(395, 33)


school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0     GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1     GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2     GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3     GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4     GP   F   16       U     GT3       T  ...     2       5        4   6  10  10

[5 rows x 33 columns]

In [ ]:
df.describe(exclude='number')

school  sex address famsize  ... nursery higher internet romantic
count     395  395     395     395  ...     395    395      395      395
unique      2    2       2       2  ...       2      2        2        2
top        GP    F       U     GT3  ...     yes    yes      yes       no
freq      349  208     307     281  ...     314    375      329      263

[4 rows x 17 columns]

In [ ]:
df.dtypes

school        object
sex           object
age            int64
address       object
famsize       object
Pstatus       object
Medu           int64
Fedu           int64
Mjob          object
Fjob          object
reason        object
guardian      object
traveltime     int64
studytime      int64
failures       int64
schoolsup     object
famsup        object
paid          object
activities    object
nursery       object
higher        object
internet      object
romantic      object
famrel         int64
freetime       int64
goout          int64
Dalc           int64
Walc           int64
health         int64
absences       int64
G1             int64
G2             int64
G3             int64
dtype: object

### 2) Use Chi^2 tests and `stats.chi2_contingency()` to identify:
 - Two pairs of variables that are dependent (are associated with one another).
 - Two pairs of variables that are independent (have no significant relationship).

Does it make intuitive sense why the variables in these pairs might or might not show a relationship to one another? 


In [ ]:
df.replace(to_replace=['no', 'yes'], value=[0, 1])

school sex  age address famsize Pstatus  ...  Walc  health absences  G1  G2  G3
0       GP   F   18       U     GT3       A  ...     1       3        6   5   6   6
1       GP   F   17       U     GT3       T  ...     1       3        4   5   5   6
2       GP   F   15       U     LE3       T  ...     3       3       10   7   8  10
3       GP   F   15       U     GT3       T  ...     1       5        2  15  14  15
4       GP   F   16       U     GT3       T  ...     2       5        4   6  10  10
..     ...  ..  ...     ...     ...     ...  ...   ...     ...      ...  ..  ..  ..
390     MS   M   20       U     LE3       A  ...     5       4       11   9   9   9
391     MS   M   17       U     LE3       T  ...     4       2        3  14  16  16
392     MS   M   21       R     GT3       T  ...     3       3        3  10   8   7
393     MS   M   18       R     LE3       T  ...     4       5        0  11  12  10
394     MS   M   19       U     LE3       T  ...     3       5        5   8   9   9

[395 rows x 33 columns]

example of dependence

In [ ]:
contingency = pd.crosstab(df['activities'], df['freetime'], margins=True)

contingency

freetime     1   2    3    4   5  All
activities                           
no          13  33   76   57  15  194
yes          6  31   81   58  25  201
All         19  64  157  115  40  395

example of independence

In [ ]:
contingency2 = pd.crosstab(df['sex'], df['absences'], margins=True)

contingency2

absences    0  1   2  3   4  5   6  7   8  ...  26  28  30  38  40  54  56  75  All
sex                                        ...                                     
F          63  2  32  2  27  2  19  5   8  ...   1   1   0   0   1   1   1   1  208
M          52  1  33  6  26  3  12  2  14  ...   0   0   1   1   0   0   0   0  187
All       115  3  65  8  53  5  31  7  22  ...   1   1   1   1   1   1   1   1  395

[3 rows x 35 columns]

In [ ]:
observed = pd.crosstab(df['sex'], df['freetime'])
observed

freetime   1   2   3   4   5
sex                         
F         15  38  93  54   8
M          4  26  64  61  32

### 3) Use NumPy to perform your own chi^2 test "from scratch" 

Pick any of the chi^2 tests that you ran in part 2 and try them on your own. You should get the same results that Scipy got for all four values returned from `chi2_contingency()`

In [ ]:
row_totals = df['freetime'].value_counts().sort_index().values

row_totals

array([ 19,  64, 157, 115,  40])

In [ ]:
col_totals = df['sex'].value_counts().sort_index().values

col_totals

array([208, 187])

In [ ]:
sample_size = df.shape[0]

sample_size

395

In [ ]:
expected = np.array([[0.0,0.0,0.0,0.0,0.0],
                     [0.0,0.0,0.0,0.0,0.0]])

for row_position, col in enumerate(col_totals):
  for column_position, row in enumerate(row_totals):
    expected_value = (row*col) / sample_size
    expected[row_position][column_position] = expected_value

expected

array([[10.00506329, 33.70126582, 82.67341772, 60.55696203, 21.06329114],
       [ 8.99493671, 30.29873418, 74.32658228, 54.44303797, 18.93670886]])

In [ ]:
chi2 = ((observed - expected)**2 / expected).sum().sum()

chi2

27.763212100392774

the p value in this case is o and the chi^2 stat is 27.763212100392774

https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html

Degrees of Freedom is different in the 2-variable chi^2 test (test for independence)

1-sample (goodness of fit), DOF = # categories-1

2-sample (test for independence), DOF = (# rows_crosstab-1)*(# cols_crosstab-1)

(4-1)*(2-1) == 3

DOF: 3

In [ ]:
chi2, p_value, dof, expected = stats.chi2_contingency(observed)

print("chi2 statistic", chi2)
print("p value", p_value)
print("degrees of freedom",dof)
print("expected frequencies table \n", expected)

chi2 statistic 27.76321210039277
p value 1.3929800482563467e-05
degrees of freedom 4
expected frequencies table 
 [[10.00506329 33.70126582 82.67341772 60.55696203 21.06329114]
 [ 8.99493671 30.29873418 74.32658228 54.44303797 18.93670886]]


## Stretch goals:

### 1. Refactor your code so it is elegant, readable, and holds reusable code in functions.

In [ ]:
# YOUR WORK HERE



### 2. Check For Understanding - Study and write your own explanations/definitions for these topics:

- What is a sample "estimate" in statistics?

- What are hypothesis test? How are they useful?

- What is a "null hypothesis?"

- What is a p-value and what does it represent?

- What does it mean for something to be "statistically significant?"

- What is a test statistic and how does it relate to a p-value?

- What are "degrees of freedom" and how are they calculated in a 1-sample, 2-sample, and chi^2 test for independence?

## Resources

- [Interactive visualize the Chi-Squared test](https://homepage.divms.uiowa.edu/~mbognar/applets/chisq.html)
- [Calculation of Chi-Squared test statistic](https://en.wikipedia.org/wiki/Pearson%27s_chi-squared_test)
- [Visualization of a confidence interval generated by R code](https://commons.wikimedia.org/wiki/File:Confidence-interval.svg)
- [Expected value of a squared standard normal](https://math.stackexchange.com/questions/264061/expected-value-calculation-for-squared-normal-distribution) (it's 1 - which is why the expected value of a Chi-Squared with $n$ degrees of freedom is $n$, as it's the sum of $n$ squared standard normals)